src/msw/spinctrl.cpp

```diff
@@ -348,57 +348,56 @@ bool wxSpinCtrl::Create(wxWindow *parent,
     if (!m_hasFont)
         SetFont(GetDefaultAttributes().font);
 
+    // If the initial text value is actually a number, it overrides the
+    // "initial" argument specified later.
+    long initialFromText;
+    if ( value.ToLong(&initialFromText) )
+        initial = initialFromText;
+
+    // Set the range in the native control: notice that we must do it before
+    // calling SetValue() to use the correct validity checks for the initial
+    // value.
+    SetRange(min, max);
+    SetValue(initial);
+
+    // Also set the text part of the control if it was specified independently
+    // but don't generate an event for this, it would be unexpected.
+    m_blockEvent = true;
+    if ( !value.empty() )
+        SetValue(value);
+    m_blockEvent = false;
+
     // Finally deal with the size: notice that this can only be done now both
     // windows are created and the text one is set up as buddy because
     // UDM_SETBUDDY changes its size using some unknown algorithm, so setting
-    // the sizes earlier is useless.
-    const int bestSpinWidth = wxSpinButton::DoGetBestSize().x;
-    const int effectiveSpinWidth = bestSpinWidth - GetOverlap();
+    // the sizes earlier is useless. Do it after setting the range and the base
+    // because DoGetBestSize() uses them.
+    wxSize bestSize = DoGetBestSize();
     wxSize sizeCtrl(size);
     if ( sizeCtrl.x <= 0 )
     {
-        // DEFAULT_ITEM_WIDTH is the default width for the text control
-        sizeCtrl.x = FromDIP(DEFAULT_ITEM_WIDTH) + effectiveSpinWidth;
+        sizeCtrl.x = bestSize.GetWidth();
     }
-    else if ( sizeCtrl.x <= effectiveSpinWidth )
+    else if ( sizeCtrl.x <= bestSize.GetWidth() )
     {
         wxLogDebug(wxS("wxSpinCtrl \"%s\": initial width %d is too small, ")
                    wxS("at least %d pixels needed."),
-                   name, size.x, effectiveSpinWidth);
+                   name, size.x, bestSize.GetWidth());
     }
 
     // adjust an invalid height for text control
     if ( sizeCtrl.y <= 0 )
     {
-        int cx, cy;
-        wxGetCharSize(GetHWND(), &cx, &cy, GetFont());
-
-        sizeCtrl.y = EDIT_HEIGHT_FROM_CHAR_HEIGHT(cy);
+        sizeCtrl.y = bestSize.GetHeight();
     }
 
     // This will call our DoMoveWindow() and lay out the windows correctly.
     SetInitialSize(sizeCtrl);
 
-    (void)::ShowWindow(GetBuddyHwnd(), SW_SHOW);
+    // Fix the min size.
```

Comment:
Can't we avoid calling `SetInitialSize()` here? It looks wrong to call it and then undo part of what it does...

If we really can't, the comment should explain better what's going on:
```suggestion
    // SetInitialSize() also sets the min size, but it should only be set to
    // the value specified by user, not the one computed by us, so override
    // the min size set just above.
```